core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/InternalAvg.java

```diff
@@ -91,9 +91,22 @@ public String getWriteableName() {
     public InternalAvg doReduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
         long count = 0;
         double sum = 0;
+        double compensation = 0;
+        // Compute the sum of double values with Kahan summation algorithm which is more
+        // accurate than naive summation.
         for (InternalAggregation aggregation : aggregations) {
-            count += ((InternalAvg) aggregation).count;
-            sum += ((InternalAvg) aggregation).sum;
+            InternalAvg avg = (InternalAvg) aggregation;
+            count += avg.count;
+            if (Double.isNaN(sum) == false) {
```

Comment:
I would remove this if statement to keep things simple?

Reply:
Make sense.